File: spark/wallet/coop_exit.go
package wallet

import (
	"bytes"
	"context"
	"fmt"
	"time"

	"github.com/btcsuite/btcd/wire"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/google/uuid"
	"google.golang.org/protobuf/types/known/timestamppb"

	"github.com/lightsparkdev/spark"
	"github.com/lightsparkdev/spark/common"
	pb "github.com/lightsparkdev/spark/proto/spark"
	"github.com/lightsparkdev/spark/so/objects"
)

// Leaf to represent input for GetConnectorRefundSignatures.
// This should probably be combined with some other input struct
// we do for transfers.
type Leaf struct {
	LeafID         string
	OutPoint       *wire.OutPoint
	SigningPubKey  *secp256k1.PublicKey
	RefundTimeLock uint32
	AmountSats     int64
	TreeNode       *pb.TreeNode
}

// GetConnectorRefundSignatures asks the coordinator to sign refund
// transactions for leaves, spending connector outputs.
func GetConnectorRefundSignatures(
	ctx context.Context,
	config *Config,
	leaves []LeafKeyTweak,
	exitTxid []byte,
	connectorOutputs []*wire.OutPoint,
	receiverPubKey *secp256k1.PublicKey,
	expiryTime time.Time,
) (*pb.Transfer, map[string][]byte, error) {
	transfer, signaturesMap, err := signCoopExitRefunds(
		ctx, config, leaves, exitTxid, connectorOutputs, receiverPubKey, expiryTime,
	)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to sign refund transactions: %v", err)
	}

	transfer, err = SendTransferTweakKey(ctx, config, transfer, leaves, signaturesMap)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to send transfer: %v", err)
	}

	return transfer, signaturesMap, nil
}

func createConnectorRefundTransactionSigningJob(
	leafID string,
	signingPubkey []byte,
	nonce *objects.SigningNonce,
	refundTx *wire.MsgTx,
) (*pb.LeafRefundTxSigningJob, error) {
	var refundBuf bytes.Buffer
	err := refundTx.Serialize(&refundBuf)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize refund tx: %v", err)
	}
	rawTx := refundBuf.Bytes()
	// TODO(alec): we don't handle errors for this elsewhere, should we here?
	refundNonceCommitmentProto, _ := nonce.SigningCommitment().MarshalProto()

	return &pb.LeafRefundTxSigningJob{
		LeafId: leafID,
		RefundTxSigningJob: &pb.SigningJob{
			SigningPublicKey:       signingPubkey,
			RawTx:                  rawTx,
			SigningNonceCommitment: refundNonceCommitmentProto,
		},
	}, nil
}

func signCoopExitRefunds(
	ctx context.Context,
	config *Config,
	leaves []LeafKeyTweak,
	exitTxid []byte,
	connectorOutputs []*wire.OutPoint,
	receiverPubKey *secp256k1.PublicKey,
	expiryTime time.Time,
) (*pb.Transfer, map[string][]byte, error) {
	if len(leaves) != len(connectorOutputs) {
		return nil, nil, fmt.Errorf("number of leaves and connector outputs must match")
	}
	signingJobs := make([]*pb.LeafRefundTxSigningJob, 0)
	leafDataMap := make(map[string]*LeafRefundSigningData)
	for i, leaf := range leaves {
		connectorOutput := connectorOutputs[i]

		if leaf.Leaf == nil {
			return nil, nil, fmt.Errorf("leaf at index %d has nil Leaf field", i)
		}
		if leaf.Leaf.RefundTx == nil {
			return nil, nil, fmt.Errorf("leaf at index %d has nil RefundTx field", i)
		}

		currentRefundTx, err := common.TxFromRawTxBytes(leaf.Leaf.RefundTx)
		if err != nil {
			return nil, nil, fmt.Errorf("failed to parse refund tx: %v", err)
		}
		sequence, err := spark.NextSequence(currentRefundTx.TxIn[0].Sequence)
		if err != nil {
			return nil, nil, fmt.Errorf("failed to get next sequence: %v", err)
		}
		refundTx, err := createConnectorRefundTransaction(
			sequence, &currentRefundTx.TxIn[0].PreviousOutPoint, connectorOutput, int64(leaf.Leaf.Value), receiverPubKey,
		)
		if err != nil {
			return nil, nil, fmt.Errorf("failed to create refund transaction: %v", err)
		}
		nonce, _ := objects.RandomSigningNonce()
		signingPrivKey := secp256k1.PrivKeyFromBytes(leaf.SigningPrivKey)
		signingPubKey := signingPrivKey.PubKey()
		signingJob, err := createConnectorRefundTransactionSigningJob(
			leaf.Leaf.Id, signingPubKey.SerializeCompressed(), nonce, refundTx,
		)
		if err != nil {
			return nil, nil, fmt.Errorf("failed to create signing job: %v", err)
		}
		signingJobs = append(signingJobs, signingJob)

		tx, _ := common.TxFromRawTxBytes(leaf.Leaf.NodeTx)

		leafDataMap[leaf.Leaf.Id] = &LeafRefundSigningData{
			SigningPrivKey: signingPrivKey,
			RefundTx:       refundTx,
			Nonce:          nonce,
			Tx:             tx,
			Vout:           int(leaf.Leaf.Vout),
		}
	}

	sparkConn, err := common.NewGRPCConnectionWithTestTLS(config.CoodinatorAddress(), nil)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to create grpc connection: %v", err)
	}
	defer sparkConn.Close()
	sparkClient := pb.NewSparkServiceClient(sparkConn)
	token, err := AuthenticateWithConnection(ctx, config, sparkConn)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to authenticate with coordinator: %v", err)
	}
	tmpCtx := ContextWithToken(ctx, token)
	transferID, err := uuid.NewV7()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to generate transfer id: %v", err)
	}
	exitID, err := uuid.NewV7()
	if err != nil {
		return nil, nil, fmt.Errorf("failed to generate exit id: %v", err)
	}
	response, err := sparkClient.CooperativeExit(tmpCtx, &pb.CooperativeExitRequest{
		Transfer: &pb.StartTransferRequest{
			TransferId:                transferID.String(),
			LeavesToSend:              signingJobs,
			OwnerIdentityPublicKey:    config.IdentityPublicKey(),
			ReceiverIdentityPublicKey: receiverPubKey.SerializeCompressed(),
			ExpiryTime:                timestamppb.New(expiryTime),
		},
		ExitId:   exitID.String(),
		ExitTxid: exitTxid,
	})
	if err != nil {
		return nil, nil, fmt.Errorf("failed to initiate cooperative exit: %v", err)
	}
	signatures, err := signRefunds(config, leafDataMap, response.SigningResults, nil)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to sign refund transactions: %v", err)
	}

	signaturesMap := make(map[string][]byte)
	for _, signature := range signatures {
		signaturesMap[signature.NodeId] = signature.RefundTxSignature
	}

	return response.Transfer, signaturesMap, nil
}

File: sdks/js/packages/spark-sdk/src/services/coop-exit.ts
import { Transaction } from "@scure/btc-signer";
import { TransactionInput } from "@scure/btc-signer/psbt";
import {
  CooperativeExitResponse,
  LeafRefundTxSigningJob,
  Transfer,
} from "../proto/spark.js";
import {
  getP2TRScriptFromPublicKey,
  getTxFromRawTxBytes,
} from "../utils/bitcoin.js";
import { getCrypto } from "../utils/crypto.js";
import { getNextTransactionSequence } from "../utils/transaction.js";
import { WalletConfigService } from "./config.js";
import { ConnectionManager } from "./connection.js";
import {
  BaseTransferService,
  LeafKeyTweak,
  LeafRefundSigningData,
} from "./transfer.js";
import { ValidationError, NetworkError } from "../errors/types.js";
import { Network } from "../utils/network.js";

const crypto = getCrypto();

export type GetConnectorRefundSignaturesParams = {
  leaves: LeafKeyTweak[];
  exitTxId: Uint8Array;
  connectorOutputs: TransactionInput[];
  receiverPubKey: Uint8Array;
};

export class CoopExitService extends BaseTransferService {
  constructor(
    config: WalletConfigService,
    connectionManager: ConnectionManager,
  ) {
    super(config, connectionManager);
  }

  async getConnectorRefundSignatures({
    leaves,
    exitTxId,
    connectorOutputs,
    receiverPubKey,
  }: GetConnectorRefundSignaturesParams): Promise<{
    transfer: Transfer;
    signaturesMap: Map<string, Uint8Array>;
  }> {
    const { transfer, signaturesMap } = await this.signCoopExitRefunds(
      leaves,
      exitTxId,
      connectorOutputs,
      receiverPubKey,
    );
    const transferTweak = await this.sendTransferTweakKey(
      transfer,
      leaves,
      signaturesMap,
    );

    return { transfer: transferTweak, signaturesMap };
  }

  private createConnectorRefundTransaction(
    sequence: number,
    nodeOutPoint: TransactionInput,
    connectorOutput: TransactionInput,
    amountSats: bigint,
    receiverPubKey: Uint8Array,
  ): Transaction {
    const refundTx = new Transaction();
    if (!nodeOutPoint.txid || nodeOutPoint.index === undefined) {
      throw new ValidationError("Invalid node outpoint", {
        field: "nodeOutPoint",
        value: { txid: nodeOutPoint.txid, index: nodeOutPoint.index },
        expected: "Both txid and index must be defined",
      });
    }
    refundTx.addInput({
      txid: nodeOutPoint.txid,
      index: nodeOutPoint.index,
      sequence,
    });

    refundTx.addInput(connectorOutput);
    const receiverScript = getP2TRScriptFromPublicKey(
      receiverPubKey,
      this.config.getNetwork(),
    );

    refundTx.addOutput({
      script: receiverScript,
      amount: amountSats,
    });

    return refundTx;
  }
  private async signCoopExitRefunds(
    leaves: LeafKeyTweak[],
    exitTxId: Uint8Array,
    connectorOutputs: TransactionInput[],
    receiverPubKey: Uint8Array,
  ): Promise<{ transfer: Transfer; signaturesMap: Map<string, Uint8Array> }> {
    if (leaves.length !== connectorOutputs.length) {
      throw new ValidationError(
        "Mismatch between leaves and connector outputs",
        {
          field: "leaves/connectorOutputs",
          value: {
            leavesCount: leaves.length,
            outputsCount: connectorOutputs.length,
          },
          expected: "Equal length",
        },
      );
    }

    const signingJobs: LeafRefundTxSigningJob[] = [];
    const leafDataMap: Map<string, LeafRefundSigningData> = new Map();

    for (let i = 0; i < leaves.length; i++) {
      const leaf = leaves[i];
      if (!leaf) {
        throw new ValidationError("Missing leaf", {
          field: "leaf",
          value: leaf,
          expected: "Valid leaf object",
        });
      }
      const connectorOutput = connectorOutputs[i];
      if (!connectorOutput) {
        throw new ValidationError("Missing connector output", {
          field: "connectorOutput",
          value: connectorOutput,
          expected: "Valid connector output",
        });
      }
      const currentRefundTx = getTxFromRawTxBytes(leaf.leaf.refundTx);

      const { nextSequence } = getNextTransactionSequence(
        currentRefundTx.getInput(0).sequence,
      );

      const refundTx = this.createConnectorRefundTransaction(
        nextSequence,
        currentRefundTx.getInput(0),
        connectorOutput,
        BigInt(leaf.leaf.value),
        receiverPubKey,
      );

      const signingNonceCommitment =
        await this.config.signer.getRandomSigningCommitment();
      const signingJob: LeafRefundTxSigningJob = {
        leafId: leaf.leaf.id,
        refundTxSigningJob: {
          signingPublicKey: leaf.signingPubKey,
          rawTx: refundTx.toBytes(),
          signingNonceCommitment: signingNonceCommitment,
        },
      };

      signingJobs.push(signingJob);
      const tx = getTxFromRawTxBytes(leaf.leaf.nodeTx);
      leafDataMap.set(leaf.leaf.id, {
        signingPubKey: leaf.signingPubKey,
        refundTx,
        signingNonceCommitment,
        tx,
        vout: leaf.leaf.vout,
        receivingPubkey: receiverPubKey,
      });
    }

    const sparkClient = await this.connectionManager.createSparkClient(
      this.config.getCoordinatorAddress(),
    );

    let response: CooperativeExitResponse;
    try {
      response = await sparkClient.cooperative_exit({
        transfer: {
          transferId: crypto.randomUUID(),
          leavesToSend: signingJobs,
          ownerIdentityPublicKey:
            await this.config.signer.getIdentityPublicKey(),
          receiverIdentityPublicKey: receiverPubKey,
          expiryTime:
            this.config.getNetwork() == Network.MAINNET
              ? new Date(Date.now() + 6 * 60 * 60 * 1000 + 5 * 60 * 1000)
              : new Date(Date.now() + 5 * 60 * 1000), // 6 hours 5 min for mainnet, 5 min otherwise
        },
        exitId: crypto.randomUUID(),
        exitTxid: exitTxId,
      });
    } catch (error) {
      throw new NetworkError(
        "Failed to initiate cooperative exit",
        {
          operation: "cooperative_exit",
          errorCount: 1,
          errors: error instanceof Error ? error.message : String(error),
        },
        error as Error,
      );
    }

    if (!response.transfer) {
      throw new NetworkError("Failed to initiate cooperative exit", {
        operation: "cooperative_exit",
        errors: "No transfer in response",
      });
    }

    const signatures = await this.signRefunds(
      leafDataMap,
      response.signingResults,
    );

    const signaturesMap: Map<string, Uint8Array> = new Map();
    for (const signature of signatures) {
      signaturesMap.set(signature.nodeId, signature.refundTxSignature);
    }

    return { transfer: response.transfer, signaturesMap };
  }
}
